internal/datastore/crdb/schema/indexutil.go

```diff
@@ -0,0 +1,222 @@
+package schema
+
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/authzed/spicedb/internal/datastore/common"
+	"github.com/authzed/spicedb/pkg/datastore"
+	"github.com/authzed/spicedb/pkg/spiceerrors"
+)
+
+var parsedColumnsPerIndex = map[string][]string{}
+
+func init() {
+	mustInit()
+}
+
+func mustInit() {
+	for _, idx := range crdbAllIndexes {
+		parsed, err := parseIndexColumns(idx.ColumnsSQL)
+		if err != nil {
+			panic(err)
+		}
+		parsedColumnsPerIndex[idx.Name] = parsed
+	}
+}
+
+func parseIndexColumns(columnsSQL string) ([]string, error) {
+	// Match columns within parentheses, handling both PRIMARY KEY and table_name formats
+	re := regexp.MustCompile(`\(([^)]+)\)`)
+	matches := re.FindStringSubmatch(columnsSQL)
+	if len(matches) < 2 {
+		return nil, fmt.Errorf("no columns found in parentheses in SQL: %s", columnsSQL)
+	}
+
+	// Split by comma and trim whitespace
+	columnsStr := matches[1]
+	columns := regexp.MustCompile(`\s*,\s*`).Split(columnsStr, -1)
+
+	// Trim any remaining whitespace
+	for i, col := range columns {
+		columns[i] = regexp.MustCompile(`^\s+|\s+$`).ReplaceAllString(col, "")
+	}
+
+	return columns, nil
+}
+
+func forcedIndexForFilter(filter datastore.RelationshipsFilter, indexes []common.IndexDefinition) (*common.IndexDefinition, error) {
+	// Algorithm: Find the index that has the most leading columns matching the filter.
+	// If at any point a column within the index is not in the filter, we stop checking for
+	// that index. We return the index with the most leading columns matched. Resource IDs
+	// are treated as an *immediate* stop, as prefix scanning does not work well with
+	// a following field in the index.
```

Comment:
Is this because we don't know whether a resource ID is prefixed or not at this point in the code?

Reply:
No; its because once we've seen a prefix, the rest of the index cannot be efficiently used